src/gc-alloc-profiler.cpp

```diff
@@ -0,0 +1,132 @@
+// This file is a part of Julia. License is MIT: https://julialang.org/license
+
+#include "gc-alloc-profiler.h"
+
+#include "julia_internal.h"
+#include "gc.h"
+
+#include <string>
+#include <vector>
+
+using std::string;
+using std::vector;
+
+struct RawBacktrace {
+    jl_bt_element_t *data;
+    size_t size;
+};
+
+struct RawAlloc {
+    jl_datatype_t *type_address;
+    RawBacktrace backtrace;
+    size_t size;
+};
+
+// == These structs define the global singleton profile buffer that will be used by
+// callbacks to store profile results. ==
+struct PerThreadAllocProfile {
+    vector<RawAlloc> allocs;
+};
+
+struct AllocProfile {
+    double sample_rate;
+
+    vector<PerThreadAllocProfile> per_thread_profiles;
+};
+
+struct CombinedResults {
+    vector<RawAlloc> combined_allocs;
+};
+
+// == Global variables manipulated by callbacks ==
+
+AllocProfile g_alloc_profile;
+int g_alloc_profile_enabled = false;
+CombinedResults g_combined_results; // Will live forever.
+
+// === stack stuff ===
+
+RawBacktrace get_raw_backtrace() {
+    // A single large buffer to record backtraces onto
+    static jl_bt_element_t static_bt_data[JL_MAX_BT_SIZE];
+
+    size_t bt_size = rec_backtrace(static_bt_data, JL_MAX_BT_SIZE, 2);
+
+    // Then we copy only the needed bytes out of the buffer into our profile.
+    size_t bt_bytes = bt_size * sizeof(jl_bt_element_t);
+    jl_bt_element_t *bt_data = (jl_bt_element_t*) malloc(bt_bytes);
+    memcpy(bt_data, static_bt_data, bt_bytes);
+
+    return RawBacktrace{
+        bt_data,
+        bt_size
+    };
+}
+
+// == exported interface ==
+
+extern "C" {  // Needed since the function doesn't take any arguments.
+
+JL_DLLEXPORT void jl_start_alloc_profile(double sample_rate) {
+    // We only need to do this once, the first time this is called.
+    while (g_alloc_profile.per_thread_profiles.size() < jl_n_threads) {
+        g_alloc_profile.per_thread_profiles.push_back(PerThreadAllocProfile{});
+    }
+
+    g_alloc_profile.sample_rate = sample_rate;
+
+    g_alloc_profile_enabled = true;
+}
+
+JL_DLLEXPORT struct RawAllocResults jl_fetch_alloc_profile() {
```

Comment:
Could we just return a Julia tuple?

Reply:
yeah I guess so 🤔 🤷‍♂️

Reply:
Seems reasonable to me. Is that more idiomatic? What's pushing you one way or the other?